src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardSetAddOrRemoveByContains.cs

```diff
@@ -0,0 +1,347 @@
+                    !AreInvocationsOnSameInstance(containsInvocation, addOrRemoveInvocation) ||
+                    !AreInvocationArgumentsEqual(containsInvocation, addOrRemoveInvocation))
+                {
+                    return;
+                }
+
+                using var locations = ArrayBuilder<Location>.GetInstance(2);
+                locations.Add(conditional.Syntax.GetLocation());
+                locations.Add(addOrRemoveInvocation.Syntax.Parent!.GetLocation());
+
+                // Build custom format instead of CSharpShortErrorMessageFormat/VisualBasicShortErrorMessageFormat to prevent unhelpful messages for VB.
+                var symbolDisplayFormat = SymbolDisplayFormat.MinimallyQualifiedFormat
+                    .WithParameterOptions(SymbolDisplayParameterOptions.IncludeType)
+                    .WithGenericsOptions(SymbolDisplayGenericsOptions.None)
+                    .WithMemberOptions(SymbolDisplayMemberOptions.IncludeParameters | SymbolDisplayMemberOptions.IncludeContainingType)
+                    .WithKindOptions(SymbolDisplayKindOptions.None);
+
+                context.ReportDiagnostic(containsInvocation.CreateDiagnostic(
+                    Rule,
+                    additionalLocations: locations.ToImmutable(),
+                    properties: null,
+                    addOrRemoveInvocation.TargetMethod.ToDisplayString(symbolDisplayFormat),
+                    containsInvocation.TargetMethod.ToDisplayString(symbolDisplayFormat)));
+            }
+        }
+
+        private static bool AreInvocationsOnSameInstance(IInvocationOperation invocation1, IInvocationOperation invocation2)
+        {
+            return (invocation1.Instance, invocation2.Instance) switch
+            {
+                (IFieldReferenceOperation fieldRef1, IFieldReferenceOperation fieldRef2) => fieldRef1.Member == fieldRef2.Member,
+                (IPropertyReferenceOperation propRef1, IPropertyReferenceOperation propRef2) => propRef1.Member == propRef2.Member,
+                (IParameterReferenceOperation paramRef1, IParameterReferenceOperation paramRef2) => paramRef1.Parameter == paramRef2.Parameter,
+                (ILocalReferenceOperation localRef1, ILocalReferenceOperation localRef2) => localRef1.Local == localRef2.Local,
+                _ => false,
+            };
+        }
+
+        // Checks if invocation argument values are equal
+        //   1. Not equal: Contains(item) != Add(otherItem), Contains("const") != Add("other const")
+        //   2. Identical: Contains(item) == Add(item), Contains("const") == Add("const")
+        private static bool AreInvocationArgumentsEqual(IInvocationOperation invocation1, IInvocationOperation invocation2)
+        {
+            return invocation1.Arguments
+                .Zip(invocation2.Arguments, (a1, a2) => IsArgumentValueEqual(a1.Value, a2.Value))
+                .All(argumentsEqual => argumentsEqual);
```

Comment:
Can't we just compare the 1st arguments? I think all involved methods has only 1 argument
```suggestion
            return IsArgumentValueEqual(invocation1.Arguments[0].Value, invocation2.Arguments[0].Value);
```